File: Sources/Interfaces/Digital Interfaces/Keyer/KeyerInterface.m
//
//  KeyerInterface.m
//  cocoaModem 2.0
//
//  Created by Kok Chen on 2/11/08.
//  Copyright 2008 Kok Chen, W7AY. All rights reserved.
//
//  The KeyerInterface class was ported to Swift (see KeyerInterface.swift).
//  This file is retained in the build ONLY to keep the two global C helper
//  functions as real C-linkage symbols for Router.swift / MicroKeyer.swift.
//

#import "KeyerInterface.h"
#import "RouterCommands.h"

#include <unistd.h>
#include <fcntl.h>
#include <string.h>

//  Open a pair of ports to the parent ports for read and write to a given type of connection
//	if one of the result pointers is nil, no open is made to that file descriptor
//	return -1 in all non-nil file descriptors of cannot open connection

void obtainRouterPorts( int *readFileDescriptor, int *writeFileDescriptor, int type, int parentReadFileDescriptor, int parentWriteFileDescriptor )
{
	char path[26], string[20], request[1] = { type } ;

	write( parentWriteFileDescriptor, request, 1 ) ;
	if (  read( parentReadFileDescriptor, string, 20 ) > 0 ) {

		if ( writeFileDescriptor != NULL ) {
			if ( string[0] == 0 ) *writeFileDescriptor = -1 ;
			else {
				strcpy( path, string ) ;
				strcat( path, "Write" ) ;
				*writeFileDescriptor = open( path, O_WRONLY ) ;
			}
		}
		if ( readFileDescriptor ) {
			if ( string[0] == 0 ) *readFileDescriptor = -1 ;
			else {
				strcpy( path, string ) ;
				strcat( path, "Read" ) ;
				*readFileDescriptor = open( path, O_RDONLY ) ;
			}
		}
	}
}

//  v0.89
//	Use OPenKEYER instead of OPENMICROKEYER, OPENCWKEYER or OPENDIGIKEYER
void obtainKeyerPortsFromKeyerID( int *readFileDescriptor, int *writeFileDescriptor, char *keyerID, int parentReadFileDescriptor, int parentWriteFileDescriptor )
{
	char path[72], string[64], request[32] ;
	int i, n ;

	request[0]= OPENKEYER ;
	n = strlen( keyerID ) ;
	for ( i = 0; i < n; i++ ) request[i+1] = keyerID[i] ;
	request[i+1] = 0 ;
	write( parentWriteFileDescriptor, request, n+2 ) ;

	n = read( parentReadFileDescriptor, string, 63 );
	if ( n > 0 ) {
		if ( writeFileDescriptor != NULL ) {
			if ( string[0] == 0 ) *writeFileDescriptor = -1 ;
			else {
				strcpy( path, string ) ;
				strcat( path, "Write" ) ;
				*writeFileDescriptor = open( path, O_WRONLY ) ;
			}
		}
		if ( readFileDescriptor ) {
			if ( string[0] == 0 ) *readFileDescriptor = -1 ;
			else {
				strcpy( path, string ) ;
				strcat( path, "Read" ) ;
				*readFileDescriptor = open( path, O_RDONLY ) ;
			}
		}
	}
}

File: Sources/Interfaces/Digital Interfaces/Keyer/KeyerInterface.swift
//
//  KeyerInterface.swift
//  cocoaModem 2.0
//
//  Created by Kok Chen on 2/11/08.
//  Swift port of KeyerInterface.m (the class only).
//  Copyright 2008 Kok Chen, W7AY. All rights reserved.
//
//  NOTE: the two global C helpers obtainRouterPorts()/obtainKeyerPortsFromKeyerID()
//  remain as C in KeyerInterface.m / KeyerInterface.h so that Router.swift,
//  MicroKeyer.swift and any Obj-C caller keep a stable C linkage symbol.
//

import Cocoa

//  Per-keyer cache carried by PTTHub / FSKHub (was the C struct MicroHamKeyerCache).
struct MicroHamKeyerCache {
	var keyer: MicroKeyer? = nil
	var controlPort: Int32 = 0
	//  FSK
	var fskPort: Int32 = 0
	var flagsPort: Int32 = 0
	var currentBaudConstant: Int32 = 0
	var currentTxInvert: Int32 = 0
	var currentStopIndex: Int32 = 0
	//  PTT
	var pttPort: Int32 = 0
}

//  common base class for PTTHub and FSKHub to provide linkage to Router (µH Router)

@objc(KeyerInterface)
class KeyerInterface: AppleScriptSupport {

	//  µH Router -- base ivars reached by PTTHub / FSKHub subclasses -> internal
	internal var selectedMicroKeyer: MicroKeyer?
	internal var router: Router?
	internal var microKeyerCache = [MicroHamKeyerCache]( repeating: MicroHamKeyerCache(), count: 16 )
	internal var activeKeyers: Int32 = 0

	override init()
	{
		super.init()
		//  0 - 1 bit, 1 = 1.5 stop bits, 2 = 2 stop bits
		var cache = MicroHamKeyerCache()
		cache.currentStopIndex = 1
		microKeyerCache = [MicroHamKeyerCache]( repeating: cache, count: 16 )
		activeKeyers = 0
		router = nil
		selectedMicroKeyer = nil
	}

	//	v0.87
	//  mode =	0 - CW
	//			1 - Voice
	//			2 - FSK
	//			3 - Digital
	@objc(setKeyerMode:controlPort:)
	func setKeyerMode( _ mode: Int32, controlPort port: Int32 )
	{
		if port <= 0 { return }

		var control: [UInt8] = [ 0x0f, 0x02, 0x00, 0x8f ]
		control[2] = UInt8( truncatingIfNeeded: mode )
		_ = control.withUnsafeBufferPointer { write( port, $0.baseAddress, 4 ) }
	}

	@objc func digiKeyerControlPort() -> Int32
	{
		//  override by sub class
		return 0
	}

	@objc func microKeyerControlPort() -> Int32
	{
		//  override by sub class
		return 0
	}

	@objc func cwKeyerControlPort() -> Int32
	{
		//  override by sub class
		return 0
	}
}
